### PR TITLE
[BO] Verify stock if needed before updating to delivered

### PR DESCRIPTION
**Pull request by Ginkosama**
There is a stock issue occurring if you follow these steps :

- On a Prestashop with Advanced Stock Management and Back-order allowed
- Create two warehouses A and B
- Create a product P that can be stored in both warehouses
- Create an order with 2 P
- Update the order status to payment accepted ( the warehouse A is the default one, so it's real quantity goes to -2 )
- When your supplier deliver you the 2 P, you make a mistake and increase the quantity in the warehouse B
- Update the order status to delivered

The warehouse A usable and real quantity are now at 0, and the warehouse B usable and real quantity are at 2, resulting in an overall quantity of 2, which is wrong.

One can think that the physical quantity of the warehouse A should be at -2, resulting in an overall quantity of 0. But that is just wrong, we can't have negative stock value in the physical world. 0 is the correct value here.

The problem is that when removeProduct() is called, there is no check for success, it is just assumed that it was done correctly.

As there is not enough quantity in warehouse A to deliver the products, it should not let us update the status to delivered.

This commit fix that by checking if the quantity can be removed, if necessary, before allowing the status update to delivered.

In this case an error will be displayed and a transfer of 2 P from warehouse B to A will be necessary before being able to deliver the product.

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop / 1.7.5.x / 1.6.1.x
| Description?  | See this PR message
| Type?         | bug fix / improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | 
| How to test?  | See this PR message

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/10919)

### CODE REVIEW

File: controllers/admin/AdminOrdersController.php [hunk 1]
```diff
@@ -388,35 +388,39 @@ public function processBulkUpdateOrderStatus()
                             if ($current_order_state->id == $order_state->id) {
                                 $this->errors[] = $this->trans('Order #%d has already been assigned this status.', array('#%d' => $id_order), 'Admin.Orderscustomers.Notification');
                             } else {
-                                $history = new OrderHistory();
-                                $history->id_order = $order->id;
-                                $history->id_employee = (int) $this->context->employee->id;
+                                if(!$this->canRemoveQuantitiesIfNeeded($order->getProductsDetail(), $current_order_state, $order_state)) {
+                                    $this->errors[] = $this->trans('Some quantities from order #%d cannot be removed from the stock.', array('#%d' => $id_order), 'Admin.Orderscustomers.Notification');
```
Comment (LouiseBonnard) [inline controllers/admin/AdminOrdersController.php]: Would it be possible to bring an element of solution to help the merchants here? What is the behavior expected?

Comment (Ginkosama) [reply controllers/admin/AdminOrdersController.php]: Yes this message can be much more explicit, i will make a commit to improve it

Comment (Ginkosama) [reply controllers/admin/AdminOrdersController.php]: The behavior expected is to check the quantities from the order in question, as some are insufficient  to remove them from the stock and pass the order to a delivered status.

Comment (LouiseBonnard) [reply controllers/admin/AdminOrdersController.php]: Great, thanks @Ginkosama, I'll wait for your suggestion and then we will be able to see how to explicit the wording to better help merchants. :-)

**Comment by PierreRambaud** [comment]
The order page has been migrated in the latest release. 
We are sorry for the delay but the change is no more relevant.

Kind regards

**Pull request closed without merging**